### ISSUE TITLE
How is training stabilized?

### ISSUE THREAD

**Issue by Cascol-Chen** [test-time-training/ttt-lm-pytorch#21]
Thanks for sharing such a great work. However, since $\theta_K$ and $\theta_V$ share the same dimension, and $f(\cdot;\cdot)$ is simply a linear function, I have the following questions:
1. How is training stabilized? Would optimization lead to collapse where $\theta_K=\theta_V$ and $W$ actually capture nothing?
2. What does $W$ capture? Since Eqn. (4) also relies on $\theta_K$ for reconstruction, how is $W$ analyzed independently?

**Comment by karan-dalal** [comment]
There are a few reasons why collapse doesn't occur:

- We can interpret K and V as feature projections to the choice of inner model. They are learned as components of the outer loop, and are optimized against a different loss than _f_. Thus, we don't learn collapse, but rather a feature projection that will be beneficial to our outer loop loss (cross entropy next token prediction).
- There's a non-linearity after all choice of _f_  in our paper.

_W_ is the weights of _f_ (matrix for TTT-Linear, 2 matrices for TTT-MLP)

**Comment by Cascol-Chen** [comment]
Thanks for the quick response. I notice that non-linearity is ```LayerNorm``` in the implementation. However, the reconstruction_target computed [here](https://github.com/test-time-training/ttt-lm-pytorch/blob/main/ttt.py#L970) does not use ```LayerNorm``` for XK. Could your provide more explanation?

**Comment by karan-dalal** [comment]
The LayerNorm is not on the reconstruction target, it's included on our choice of _f_. 

We add a residual connection to _f_ for training stability (Sec. 2.7). So _f(x) = x + LN(f_res(x))_. Since we use MSE loss, we can pre-compute the reconstruction target as the difference between the residual and label view.

The LN is [applied here](https://github.com/test-time-training/ttt-lm-pytorch/blob/cd831db10c8c9a0f6340f02da5613316a8a92b67/ttt.py#L1021) during the forward of the test view.

**Comment by Cascol-Chen** [comment]
Thanks again for your quick and thoughtful response. I believe this work will broaden our understanding of TTT/TTA, and I am hopeful that TTT/TTA will become a standard practice in deep learning.
